Implemented class Str,basic operations for string

a-moreno-a opened the pull request:
Implemented basic operator (for example: [], +, +=, = ), constructors(copy and overloads), destructor, created some simple tests in main.cpp to show all operators, ran "valgrind --leak-check=yes -v" and allocations and deallocations match.

*No really sure if std::ostream &operator<<(std::ostream &os, const Str& str) should be inline to avoid redefinition issues

a-moreno-a committed:
Implemented class Str,basic operations for string

Implemented basic operator([], +, +=, 0), constructors(copy and overloads), created some simple tests in main.cpp to show all operators, ran "valgrind --leak-check=yes -v" and allocations and deallocations match.

bodorkos-adam commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+        start = new char[size];
+        end = start;
+
+        while (*s != '\0')
```
std::strcpy?

a-moreno-a replied:
That’s right

bodorkos-adam commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+
+    char operator[](int i) const
+    {
+        return *(start + i);
```
start[i]

a-moreno-a replied:
Yeah, one translates to the other, I think I was stuck in pointer arithmetics at that point of the code

bodorkos-adam commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+        mid = end;
+    }
+
+    friend void Swap(Str& str1, Str& str2)
```
This is usually done differently to enable argument dependent lookup: [https://stackoverflow.com/questions/28130671/how-does-using-stdswap-enable-adl](https://stackoverflow.com/questions/28130671/how-does-using-stdswap-enable-adl)

maciekgajewski replied:
Indeed. But containers usually have ```swap``` method as well.

a-moreno-a replied:
I am not fully aware about how ADL works, but you meant that I should’ve put as public so it can be found using unqualified calls ?

maciekgajewski replied:
No, ```swap``` should be free function, defined in the amer namespace as the object.
The good practice is to provide ```swap``` _method_, and implement swap _function_ in its terms:

```
class MyContainer
{
// ...
void swap(MyContainer& other) { /* ... */ }
};

void swap(MyContainer& a, MyContainer& b) { a.swap(b); }
```

maciekgajewski commented on course6/string_ariel/main.cpp:
```diff
@@ -0,0 +1,40 @@
+{
+    //Constructor
+    Str s1 = "foo";
+    std::cout << s1 << std::endl;
```
It's a valid interactive test, but in the end you want it to be automated. 
You can write a simple automated test using ```assert```, or do it properly with Google Test or Boost.Test

a-moreno-a replied:
That’s true, did not have time to do a proper unit testing, and hesitated to add 3rd party libraries that cannot build for other people, at least I could’ve used assert tho

maciekgajewski replied:
You should not hesitate to use Boost.
Also, if you not use any 3rd party lib, but simple ```assert``, then it will be simple and universal. Even easier to write and use than all these couts

maciekgajewski commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+
+    Str(const Str& str)
+    {
+        int size = str.Size();
```
The content of this method is almost identical to the method above. There could be one function doing this.

maciekgajewski commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+#include <iostream>
+#include <cstring>
+
+class Str
```
Wrap it into a namespace. Get into a habit of using namespaces. All the production code on the planet is inside of namespace, so it's good to get some experience with it as well.

a-moreno-a replied:
Ok, will do on my next PR

maciekgajewski commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+        return *this;
+    }
+
+    int Size() const
```
When implementing container, stick to the standard naming and interface. This way you can write generic code that can work on different containers, including yours.
Specifically, you should write your unit test as a template method, and it should pass for both your string and ```std::string```.

a-moreno-a replied:
I tried to stay away as possible from container and some std stuff that we haven’t seen in the classes

maciekgajewski commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+    }
+};
+
+std::ostream &operator<<(std::ostream &os, const Str& str) {
```
This should be inline.
Or defined in .cc file. I think some of the bigger methods in this file should go to .cc.

maciekgajewski commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+        std::swap(str1.end, str2.end);
+    }
+
+    void Dispose()
```
Why separate method for that?

maciekgajewski commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+        delete[] start;
+    }
+
+    void Copy(const char* s)
```
It's called ```std::strcpy```, and you should use it in your constructors.
You wrote the same loop several times!

maciekgajewski commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+            Copy(s);
+        }
+
+        return *this;
```
again: two almost identical methods one after another

maciekgajewski commented on course6/string_ariel/str.h:
```diff
@@ -0,0 +1,193 @@
+        }
+
+        return *this;
+    }
```
and this is almost identical to the private concatenating constructor

a-moreno-a replied:
I am not completely sold, the concatenating constructor suits well inside the else statement (which is what the + operator does) but inside the if statement I am no really interested in allocating memory, that’s why I didnt use the constructor

maciekgajewski replied:
My point is: I see the same code twice (or even thrice). This shouldn't be. There is many ways of avoiding duplication.

maciekgajewski reviewed (changes requested):
You are missing one, very important method: c_str()

a-moreno-a commented:
Agree with all the comments I didn’t reply

a-moreno-a committed:
Fixing issues based on code review

I think is addressed most of the issues mentioned, I still need to improve the unit test, I will do it for my next PR

a-moreno-a commented on course6/string_ariel/str.h:
```diff
@@ -2,192 +2,100 @@
-    }
+        Str& operator += (const Str& rh)
+        {
+            Assign((*this + rh).start, this->size() + rh.size());
```
I know this is an extra allocatoin/deallocation but makes the code much more readable

maciekgajewski reviewed (approved):
Don't forget to implement c_str()!

a-moreno-a committed:
Trying unit test using boost

Moved the Str to a different library and this is linked to a unit test file

maciekgajewski commented on course6/string_ariel/string_tests.cc:
```diff
@@ -0,0 +1,76 @@
+{
+  std::string s1 = "123";
+  Str t1 = "123";
+  BOOST_REQUIRE(s1.size() == t1.size());
```
Use ```BOOST_REQUIRE_EQUAL``` instead. This way the error message will contain the expected and got values.

maciekgajewski commented on course6/string_ariel/string_tests.cc:
```diff
@@ -0,0 +1,76 @@
+
+BOOST_AUTO_TEST_SUITE (stringtest)
+
+BOOST_AUTO_TEST_CASE (constructor_tests)
```
Add test for default constructed string: make sure ```s.size() == 0```, ```std::strlen(s.c_str()) == 0```, ```s == ""``` etc

maciekgajewski commented on course6/string_ariel/CMakeLists.txt:
```diff
@@ -1,6 +1,12 @@
 
 project(string_ariel)
+
+add_subdirectory(str)
```
You should add ``find_package(Boost REQUIRED ...)```, and then use variables set by it. Otherwise it is impossible to compile the code if Boost is installed outside of standard location

maciekgajewski replied:
You need o set C++ standard to 14, otherwise the code won't compile on some compilers

maciekgajewski reviewed (changes requested):
Nice! Few small inefficiencies and (at least) one potentially crashing bug, but good!
Tests for empty string are missing, please, add them.
